Hoststatus remove stats creation.

jrushford opened the pull request:
Modifications to the traffic_ctl host status command and HostStatus

- modifies traffic_ctl to use JSON-RPC to query host statuses directly from the HostStatus singleton
- removes the persistent stats creation in HostStatus
- adds in a new persistent mechanism for HostStatus

Host status will no-longer create persistent stats so now when a host is marked up or down using **traffic_ctl,** the entire set of HostStatus records will be dumped from memory to var/trafficserver/host_records.yaml so that when ATS restarts, the records are re-loaded.  stats_over_http will no-longer show **"proxy.process.host_status"** stats and are no-longer seen with **traffic_ctl metric match**.

**traffic_ctl host status** without any other arguments will list all host status records replacing the metric match usage to list all host records.

bneradt commented:
[approve ci rocky]

bneradt commented:
[approve ci rocky]

brbzull0 commented:
https://github.com/apache/trafficserver/pull/8694 should fix the autest. Once it's merged into 10-Dev you can rebase.

jrushford commented:
[approve ci autest]

bneradt commented:
[approve ci autest]

bneradt commented:
[approve ci fedora]

bneradt commented:
[approve ci autest]

brbzull0 commented on src/traffic_server/traffic_server.cc:
```diff
@@ -2162,7 +2162,6 @@ main(int /* argc ATS_UNUSED */, const char **argv)
     RecProcessStart();
     initCacheControl();
     IpAllow::startup();
-    HostStatus::instance().loadHostStatusFromStats();
```
This was key I think, it was creating the first  `HostStatus` instance, without this the ctor will not be called so no handlers registered.  You should probably need a `init()` function to register the handlers.

jrushford commented:
[approve ci Fedora]

jrushford commented:
[approve ci fedora]

jrushford commented:
@brbzull0 I've added in the persistence.  You'll see that I added back in the HostStatus call in traffic_server.cc with the new HostStatus::loadFromPersistentStore() function

jrushford commented:
[approve ci clang-analyzer

jrushford commented:
[approve ci clang-analyzer]

jrushford commented:
[approve ci fedora]

brbzull0 commented:
@jrushford this looks good, just have a question.
Do you think you will be injectingthe `host_records.yaml` into traffic_server through the JSON-RPC node?(by using traffic_ctl or any other tool). If this is the plan at some point maybe the field names in the rpc request and the dump `host_records.yaml` file should be the same in order to re-use them.

jrushford commented:
> @jrushford this looks good, just have a question.
> Do you think you will be injectingthe `host_records.yaml` into traffic_server through the JSON-RPC node?(by using traffic_ctl or any other tool). If this is the plan at some point maybe the field names in the rpc request and the dump `host_records.yaml` file should be the same in order to re-use them.

@brbzull0 I didn't have any plans to use the JSON-RPC node to inject them but, that could be useful. In this PR, the host_records.yaml is loaded at startup by trafficserver.  However, I'll take a look and make sure that the names are the same, someone might have a use case for the injection.

ywkaras commented on proxy/HostStatus.h:
```diff
@@ -168,6 +175,35 @@ struct HostStatRec {
 };
 
+struct HostStatusSync;
+typedef int (HostStatusSync::*HostStatusSyncHandler)(int, void *);
```
Don't need this.

jrushford replied:
changed

ywkaras commented on proxy/HostStatus.h:
```diff
@@ -168,6 +175,35 @@ struct HostStatRec {
+  {
+    getHostStatusPersistentFilePath();
+
+    SET_HANDLER((HostStatusSyncHandler)&HostStatusSync::mainEvent);
```
You should not need to cast the type of the event function anymore because of https://github.com/apache/trafficserver/pull/6996 .

It is best to not use C casts in new code.  A C cast is equivalent to reinterpret_cast and is hard to find with grep.

jrushford replied:
@ywkaras ok, it's removed.

ywkaras commented on proxy/HostStatus.h:
```diff
@@ -182,11 +218,16 @@ struct HostStatus {
   void loadRecord(std::string_view name, HostStatRec &h);
-  RecErrT getHostStat(std::string &stat_name, char *buf, unsigned int buf_len);
+  HostStatRec *getHostStatus(const std::string_view name);
+  void getAllHostStatuses(std::vector<HostStatuses> &hosts);
```
You could return a vector, it should be efficient due to Return Value Optimization.

ywkaras replied:
We should let the people who went to grad school tell us if it's wrong.

ywkaras commented on src/traffic_ctl_jsonrpc/jsonrpc/CtrlRPCRequests.h:
```diff
@@ -129,10 +139,16 @@ struct HostSetStatusRequest : shared::rpc::ClientRequest {
+
+  std::string
+  get_method() const
+  {
```
@brbzull0 it's not clear why this even needs to be a virtual function:  https://github.com/apache/trafficserver/blob/68e69e27b87576abfaa7a1bbeb8394795c522e82/include/shared/rpc/RPCRequests.h#L45

Should derived classes just set the "method" member string?

brbzull0 replied:
> it's not clear why this even needs to be a virtual function

Because of this https://github.com/apache/trafficserver/blob/68e69e27b87576abfaa7a1bbeb8394795c522e82/include/shared/rpc/yaml_codecs.h#L211
 
>Should derived classes just set the "method" member string?

yes they can, in which case `JSONRPCRequest::get_method` will return it.

I wasn't very concern about the virtual thing in here as this is used by `traffic_ctl` and `traffic_top`.

ywkaras commented on src/traffic_server/HostStatus.cc:
```diff
@@ -20,85 +20,32 @@
-  }
-}
+ts::Rv<YAML::Node> server_get_status(std::string_view const &id, YAML::Node const &params);
+ts::Rv<YAML::Node> server_set_status(std::string_view const &id, YAML::Node const &params);
```
Why reference to string_view ?

jrushford replied:
@ywkaras okay, will pass by value.

brbzull0 replied:
This is the RPC handler API:

https://github.com/apache/trafficserver/blob/68e69e27b87576abfaa7a1bbeb8394795c522e82/mgmt2/rpc/jsonrpc/JsonRPCManager.h#L64

The reason is `const&` is becase I wanted  to make it clear that this should not mutate despite the fact that you cannot change it if it was only a `std::string_view`. I always thought that this is ambiguous and if someone try to change it well, good luck. Now kind of I am inclined to remove the `const &`

ywkaras commented on doc/appendices/command-line/traffic_ctl_jsonrpc.en.rst:
```diff
@@ -366,7 +365,8 @@ endpoint.
 
    Get the current status of the specified hosts with respect to their use as targets for parent
-   selection. This returns the same information as the per host stat.
+   selection. This returns the same information as the per host record. If the HOSTNAME arguments
```
But the status subcommand is explained in line 328.  It seems like what this is saying is that the host record returns the same information as the host record, so maybe we can just delete that sentence.

jrushford replied:
ahh ok, I'll do that

jrushford replied:
@ywkaras ok, it's removed.

ywkaras commented on src/traffic_server/HostStatus.cc:
```diff
@@ -44,8 +44,8 @@ struct HostCmdInfo {
-ts::Rv<YAML::Node> server_get_status(std::string_view const &id, YAML::Node const &params);
-ts::Rv<YAML::Node> server_set_status(std::string_view const &id, YAML::Node const &params);
+ts::Rv<YAML::Node> server_get_status(std::string_view const id, YAML::Node const &params);
+ts::Rv<YAML::Node> server_set_status(std::string_view const id, YAML::Node const &params);
```
It's redundant to use 'string_view const' instead of just plain 'string_view' but harmless.

jrushford replied:
@ywkaras I'll change it.

brbzull0 replied:
> It's redundant to use 'string_view const' instead of just plain 'string_view' but harmless.

Totally, but the reason is this:

https://github.com/apache/trafficserver/blob/68e69e27b87576abfaa7a1bbeb8394795c522e82/mgmt2/rpc/jsonrpc/JsonRPCManager.h#L64

It is expected to be like this.

--
The reason is `const&` is becase I wanted  to make it clear that this should not mutate despite the fact that you cannot change it if it was only a `std::string_view`. I always thought that this is ambiguous and if someone try to change it well, good luck. Now kind of I am inclined to remove the `const &`

ywkaras commented:
Looks good to me, let's see what Damian says.

jrushford commented:
[approve ci fedora]

jrushford commented:
[approve ci fedora]

jrushford commented:
[approve ci]

jrushford commented:
@brbzull0 and @ywkaras any more changes?  Are you ok with the `const &` removal?

brbzull0 commented:
Looks good to me.
